pybamm/solvers/c_solvers/idaklu/casadi_functions.hpp

```diff
@@ -3,56 +3,106 @@
 
 #include "common.hpp"
 #include "options.hpp"
-#include "solution.hpp"
 #include <casadi/casadi.hpp>
+#include <memory>
+
+// Utility function for compressed-sparse-column (CSC) to/from
+// compressed-sparse-row (CSR) matrix representation.
+template<typename T1, typename T2>
+void csc_csr(realtype f[], T1 c[], T1 r[], realtype nf[], T2 nc[], T2 nr[], int N, int cols) {
+  int nn[cols+1];
+  int rr[N];
+  for (int i=0; i<cols+1; i++)
+    nc[i] = 0;
+
+  for (int k = 0, i = 0; i < cols+1; i++) {
+    for (int j = 0; j < r[i+1] - r[i]; j++) {
+      if (k == N)  // SUNDIALS indexing does not include the count element
+        break;
+      rr[k++] = i;
+    }
+  }
+  for (int i = 0; i < N; i++)
+    nc[c[i]+1]++;
+  for (int i = 1; i <= cols; i++)
+    nc[i] += nc[i-1];
+  for (int i = 0; i < cols+1; i++)
+    nn[i] = nc[i];
+  for (int i = 0; i < N; i++) {
+    int x = nn[c[i]]++;
+    nf[x] = f[i];
+    nr[x] = rr[i];
+  }
+}
 
 using Function = casadi::Function;
 
 class CasadiFunction
 {
 public:
   explicit CasadiFunction(const Function &f);
+  void operator()();
+  void operator()(std::vector<realtype*> inputs,
+                  std::vector<realtype*> results);
 
 public:
   std::vector<const double *> m_arg;
   std::vector<double *> m_res;
-  void operator()();
 
-private:
+//private:
```

Comment:
should this be uncommented?

Reply:
m_func is accessed for nnz and sparsity values outside of the class. i have added the private declaration back in for the other variables but made m_func public rather than provide dedicated getters (unless that approach is strongly preferred?).